core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java

```diff
@@ -211,72 +203,87 @@ private synchronized void responseDone(Either<Throwable, CoreEvent> result) {
       signalConsumerSilently(onResponseConsumer);
     }
     onResponseConsumerList.clear();
+
     tryComplete();
   }
 
   protected void tryComplete() {
-    boolean allChildrenComplete;
+    if (areAllChildrenComplete() && state.compareAndSet(STATE_RESPONSE_RECEIVED, STATE_RESPONSE_PROCESSED)) {
```

Comment:
is this on purpose that the state won't change until `areAllChildrenComplete()` is true?

Reply:
also, checking the state would be slower than iterating the child context?

Reply:
the current code already tests the children before testing the state. However, the ordering is more about logic than performance. I'm using a CAS instruction so if test is positive, the value would atomically change, and that would be had to rollback